### PR TITLE
Explode pipe + span classifier training tutorial

### PR DESCRIPTION
**Pull request by percevalw**
New `eds.explode` pipe that splits one document into multiple documents, one per span yielded by its `span_getter` parameter, each new document containing exactly that single span.
The goal is mainly to allow more flexibility when designing the batch sizes for training span based components (eg. span classifier).
- at the moment, to allow the batch to "mix" well, the user must split the documents into smaller chunks, hoping to get at most a few entities per chunk. There is also the issue of context, as a span qualifier may want to train with a context of 30 words before/after, but if the chunks are smaller than that, the model will never see the context size that it will seing after its training (ie on non chunked docs)
- so splitting along the entities (assuming the prediction of the component on a given entity doesn't depend on the fact that another entity is annotated) allows 1/ to preserve context, 2/ mix batches correctly

**Change overview**

- Added:

  - New `eds.explode` pipe that splits one document into multiple documents, one per span yielded by its `span_getter` parameter, each new document containing exactly that single span.
  - New `Training a span classifier` tutorial, and reorganized deep-learning docs
- `ScheduledOptimizer` now warns when a parameter selector does not match any parameter.

- Fixed:

  - We should now correctly support loading transformers in offline mode if they were already in huggingface's cache
  - We now support `words[-10:10]` syntax in trainable span classifier `context_getter` parameter
  - :ambulance: :warning: MAJOR FIX: Until now, `post_init` was applied **after** the instantiation of the optimizer : if the model discovered new labels, and therefore changed its parameter tensors to reflect that, these new tensors were not taken into account by the optimizer, which could likely lead to subpar performance. Now, `post_init` is applied **before** the optimizer is instantiated, so that the optimizer can correctly handle the new tensors.

- Changed:
  - It is now recommended to define optimizer groups of `ScheduledOptimizer` as a list of dicts of optim hyper-parameters, each containing a `selector` regex key, rather than as a single dict with a `selector` as keys and a dict of optim hyper-parameters as values. This allows for more flexibility in defining the optimizer groups, and is more consistent with the rest of the EDS-NLP API. This makes it easier to reference groups values from other places in config files, since their path doesn't contain a complex regex string anymore. See the updated training tutorials for more details.

## Checklist

- [x] If this PR is a bug fix, the bug is documented in the test suite.
- [x] Changes were documented in the changelog (pending section).
- [x] If necessary, changes were made to the documentation (eg new pipeline).

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -142,10 +144,13 @@ nav:
       - data/polars.md
       - data/spark.md
       - data/converters.md
+  - Training:
+      - training/loggers.md
   - Concepts:
       - concepts/pipeline.md
       - concepts/torch-component.md
       - concepts/inference.md
+      - concepts/deep-learning.md
```
Comment (aricohen93) [inline mkdocs.yml:153]: idem precedent comment. Maybe put this tutorial under the Trianing section

File: pyproject.toml [hunk 2]
```diff
@@ -23,7 +23,7 @@ dependencies = [
     # blis>1.2.0 (dependency of thinc) doesn't provide binaries for python<3.10 so we need to cap it ourself
     "blis<1.0.0; python_version<'3.9'",
     "blis<1.2.1; python_version>='3.9' and python_version<'3.10'",
-    "confit>=0.7.3",
+    "confit@git+https://github.com/aphp/confit.git",
```
Comment (aricohen93) [inline pyproject.toml:26]: not possible to leave `"confit>=0.9.0"` ?

File: docs/tutorials/training-span-classifier.md [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+pip install uv
+uv venv .venv
+source .venv/bin/activate
+uv pip install -e ".[dev]"
+```
+
+## Creating the dataset
+
+We'll use a small dataset of annotated biopsy dates. The dataset is in the [standoff format](https://brat.nlplab.org/standoff). You can use [Brat](https://brat.nlplab.org/) to visualize and edit the annotations.
+The dataset is available under [`tests/training/dataset_2`](https://github.com/aphp/edsnlp/tree/master/tests/training/dataset_2) directory of EDS-NLP's repository.
```
[the remark this answers is not in the document]
Comment (percevalw) [reply docs/tutorials/training-span-classifier.md:65]: good catch, thanks !

File: docs/tutorials/training-span-classifier.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+    # 🚀 TRAIN
+    train(
+        nlp=nlp,
+        train_data=TrainingData(
+            data=train_docs,
+            batch_size="8 spans",
+            pipe_names=["biopsy_classifier"],
+            shuffle="dataset",
+        ),
+        val_data=val_docs,
```
[the remark this answers is not in the document]
Comment (percevalw) [reply docs/tutorials/training-span-classifier.md:297]: The batch size is actually in the scorer config

File: docs/tutorials/training-span-classifier.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+    be called by `train`, so the optimizer should be defined *after*, when the model parameter tensors are
+    final. To do that, instead of instantiating the optimizer, we create a "Draft", which will be
+    instantiated inside the `train` function, once all the required parameters are set.
+
+
+!!! note "Upstream annotations at training vs inference time"
+
+    In this example, the pipeline contains the `eds.dates` component
+    but this component is *not* applied to documents *during the training*.
+
```
[the remark this answers is not in the document]
Comment (percevalw) [reply docs/tutorials/training-span-classifier.md:323]: It's not that EDS-NLP blocks it, it's just that when preprocessing the data, we don't apply by default the pipes of the nlp model. To apply them to the training samples, these must be added as preprocessing steps in the converter function of the train_data object (same can be done to the val_data).
Just as augmentations, if we split the documents, or duplicate them, or do some transformations during the training, these won't be part of the final trained pipeline, but rather are only part of the stream of training samples. Pre-annotations like automatic label annotations should work the same → put them in the train_data converter
